### PR TITLE
Encoding of an enum with all unit variants (#390)

### PR DESCRIPTION
**Pull request by segfault-magnet**
closes: #390 
### The issue
The sway compiler, since the release of Forc 16.0.0, includes an optimization with regards to enums that only have unit types for variants.

Such enums are to be encoded in only one word that contains their discriminant.

This PR also contains the necessary fixes to our Sway files thanks to @digorithm 

### The solution 
The cleanest way of looking at this is to say that enums with only unit variants will not have their unit variant encoded at all. This is preferable to wording like 'units are 0B in enums, 1 B elsewhere'. Although the 'elsewhere' part can be debated since the SDK doesn't support using Units anywhere else. This phrasing was translated into code.

Although the changes are covered by unit tests, there is an e2e test that will safeguard us against regressions in the sway compiler -- the test will only pass if the aforementioned type of enums is encoded in one word.

I've also taken the liberty of refactoring ABIEncoder, since it was unnecessarily storing the encoded function selector while it was only necessary for it to compute it.

The flow of 'construct an encoder, give it data, get the encoding back' was wrapped in an associated function which can now be called `ABIEncoder::encode(...)` since this is the only way the encoder is used.

### CODE REVIEW

**Review by adlerjohn — COMMENTED** [review]
CI failing

File: packages/fuels-core/src/abi_encoder.rs [hunk 1]
```diff
@@ -2,98 +2,152 @@ use crate::constants::{ENUM_DISCRIMINANT_WORD_WIDTH, WORD_SIZE};
...
+        (&mut output[4..]).copy_from_slice(&result[..4]);
+        output
     }
 
-    pub fn new_with_fn_selector(signature: &[u8]) -> Self {
-        Self {
-            function_selector: Self::encode_function_selector(signature),
-            encoded_args: Vec::new(),
+    /// Encodes `Token`s in `args` following the ABI specs defined
+    /// [here](https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/abi.md)
```
Comment (digorithm) [inline packages/fuels-core/src/abi_encoder.rs:28]: Do you mind using the permalink here? i.e. https://github.com/FuelLabs/fuel-specs/blob/1be31f70c757d8390f74b9e1b3beb096620553eb/specs/protocol/abi.md

**Review by digorithm — CHANGES_REQUESTED** [review]
Tiny nits, other than that LGTM. Nice refactor and test 🔥

**Review by Br1ght0ne — DISMISSED** [review]
Other than what @digorithm said, LGTM 🎉

**Review by digorithm — APPROVED** [review]
:shipit:

**Pull request merged**